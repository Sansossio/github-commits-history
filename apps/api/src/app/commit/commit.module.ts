import { Module } from "@nestjs/common";
import { CommitService } from "./service/commit.service";
import { CommitController } from "./controller/commit.controller";

@Module({
  controllers: [CommitController],
  providers: [CommitService],
})
export class CommitModule {}
