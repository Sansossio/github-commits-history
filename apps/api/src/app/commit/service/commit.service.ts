import { ForbiddenException, HttpStatus, Injectable, InternalServerErrorException, NotFoundException } from "@nestjs/common";
import { GithubService } from "@github-commits-history/github";
import { Commit, GetCommits } from "@github-commits-history/github/interfaces";
import { catchError, map, Observable, } from "rxjs";
import { GetSimplifiedCommitsResponseDTO } from "../dto/get-simplified-commits/get-simplified-commits.response.dto";
import { GetCommitRequestDTO } from "../dto/get-commit/get-commit.request.dto";

@Injectable()
export class CommitService {
  constructor(private readonly github: GithubService) {}

  getCommits(query: GetCommits): Observable<Commit[]>  {
    return this.github.getCommitsHistory(query)
      .pipe(
        catchError((err) => {
          if (err.response.status === HttpStatus.NOT_FOUND) {
            throw new NotFoundException()
          }

          if (err.response.status === HttpStatus.FORBIDDEN) {
            throw new ForbiddenException("Api limit exceeded")
          }

          throw new InternalServerErrorException()
        })
      )
  }

  getCommit(query: GetCommitRequestDTO): Observable<Commit> {
    return this.github.getCommit(query)
  }

  getSimplifiedCommits(query: GetCommits): Observable<GetSimplifiedCommitsResponseDTO[]> {
    return this.getCommits(query)
      .pipe(map((val) => val.map(item => GetSimplifiedCommitsResponseDTO.fromGithubResponse(item))))
  }
}
