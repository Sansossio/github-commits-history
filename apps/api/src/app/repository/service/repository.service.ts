import { HttpStatus, Injectable, InternalServerErrorException, NotFoundException } from "@nestjs/common";
import { GithubService } from "@github-commits-history/github";
import { Commit, GetRepository } from "@github-commits-history/github/interfaces";
import { catchError, Observable, } from "rxjs";

@Injectable()
export class RepositoryService {
  constructor(private readonly github: GithubService) {}

  getCommits(query: GetRepository): Observable<Commit[]>  {
    return this.github.getCommitsHistory(query)
      .pipe(
        catchError((err) => {
          if (err.response.status === HttpStatus.NOT_FOUND) {
            throw new NotFoundException()
          }

          throw new InternalServerErrorException()
        })
      )
  }
}
