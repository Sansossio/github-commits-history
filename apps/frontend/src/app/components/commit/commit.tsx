import { Commit } from "@github-commits-history/github/interfaces"
import { TableRow } from "@mui/material";
import TableCell from '@mui/material/TableCell';
import { useCallback } from "react";
import { timeAgoFormat } from "../../utils/date.util";

export function CommitComponent({ commit }: { commit: Commit }) {
  const handleClick = useCallback(() => {
    window.open(commit.html_url, '_blank')
  }, [])

  return (
    <>
      <TableRow
        key={commit.sha}
        sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
        className="commit-row"
        onClick={() => handleClick()}
      >
        <TableCell>{timeAgoFormat(commit.commit.author?.date)}</TableCell>
        <TableCell>{commit.commit.message}</TableCell>
        <TableCell>{commit.commit.author?.name}</TableCell>
        <TableCell>{commit.sha}</TableCell>
      </TableRow>
    </>
  )
}
