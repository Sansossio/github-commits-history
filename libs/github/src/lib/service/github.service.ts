import { HttpService } from "@nestjs/axios";
import { Inject, Injectable } from "@nestjs/common";
import { map, Observable } from "rxjs";
import { GITHUB_MODULE_TOKEN } from "../github.module-definition";
import { Commit, GetRepository, RegisterGithubModule } from "../interfaces";
import { githubConfig } from '../config/github.config';

@Injectable()
export class GithubService {
  private readonly baseUrl = this.configuration.baseUrl ?? githubConfig.baseUrl;

  constructor(
    @Inject(GITHUB_MODULE_TOKEN)
    private readonly configuration: RegisterGithubModule,

    private readonly httpService: HttpService
  ) {}

  getCommitsHistory ({ owner, repository }: GetRepository): Observable<Commit[]> {
    return this.httpService.request<Commit[]>({
      baseURL: this.baseUrl,
      url: `repos/${owner}/${repository}`
    })
      .pipe(map((val) => val.data))
  }
}
