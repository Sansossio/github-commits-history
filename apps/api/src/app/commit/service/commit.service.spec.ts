import { HttpStatus, InternalServerErrorException, NotFoundException } from '@nestjs/common'
import { firstValueFrom, of, throwError } from 'rxjs'
import { CommitService } from './commit.service'

describe('CommitService', () => {
  const githubService = {
    getCommitsHistory: jest.fn(),
  }
  const service = new CommitService(githubService as any)
  const query = {
    owner: 'test_owner',
    repository: 'repo'
  }
  
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should be defined', () => {
    expect(CommitService).toBeDefined()
  })

  describe('getCommits', () => {
    it('should match snapshot', async () => {
      const response = [
        {
          id: 1234
        },
        {
          id: 321
        }
      ]
      githubService.getCommitsHistory.mockImplementation(() => of(response))

      await expect(firstValueFrom(service.getCommits(query))).resolves.toEqual(response)
    })

    it('should throw not found when fails', async () => {
      const error: any = new Error()
      error.response = {
        status: HttpStatus.NOT_FOUND
      }
      githubService.getCommitsHistory.mockImplementation(() => throwError(() => error))

      await expect(firstValueFrom(service.getCommits(query))).rejects.toBeInstanceOf(NotFoundException)
    })

    it('should throw internal server error when is an unhandler error', async () => {
      const error: any = new Error()
      error.response = {
        status: HttpStatus.BAD_GATEWAY
      }
      githubService.getCommitsHistory.mockImplementation(() => throwError(() => error))

      await expect(firstValueFrom(service.getCommits(query))).rejects.toBeInstanceOf(InternalServerErrorException)
    })
  })

  describe('getSimplifiedCommits', () => {
    it('should match snapshot', async () => {
      const response = [
        {
          sha: 'sha1',
          commit: {
            message: 'test message',
            committer: {
              date: new Date(2021, 10, 7, 1, 0, 0, 0)
            },
          },
          author: {},
          committer: {}
        },
        {
          sha: 'sha2',
          commit: {
            message: 'test message',
            committer: {
              date: new Date(2001, 10, 7, 1, 0, 0, 0)
            },
          },
          author: {},
          committer: {}
        }
      ]
      githubService.getCommitsHistory.mockImplementation(() => of(response))

      await expect(firstValueFrom(service.getSimplifiedCommits(query))).resolves.toMatchSnapshot()
    })
  })
})
