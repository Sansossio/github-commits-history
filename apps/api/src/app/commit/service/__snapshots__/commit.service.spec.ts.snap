// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitService getSimplifiedCommits should match snapshot 1`] = `
Array [
  GetSimplifiedCommitsResponseDTO {
    "author": UserGetSimplifiedCommitsResponseDTO {
      "avatar_url": undefined,
      "id": undefined,
      "login": undefined,
      "url": undefined,
    },
    "committer": UserGetSimplifiedCommitsResponseDTO {
      "avatar_url": undefined,
      "id": undefined,
      "login": undefined,
      "url": undefined,
    },
    "date": 2021-11-07T05:00:00.000Z,
    "message": "test message",
    "sha": "sha1",
    "url": undefined,
  },
  GetSimplifiedCommitsResponseDTO {
    "author": UserGetSimplifiedCommitsResponseDTO {
      "avatar_url": undefined,
      "id": undefined,
      "login": undefined,
      "url": undefined,
    },
    "committer": UserGetSimplifiedCommitsResponseDTO {
      "avatar_url": undefined,
      "id": undefined,
      "login": undefined,
      "url": undefined,
    },
    "date": 2001-11-07T05:00:00.000Z,
    "message": "test message",
    "sha": "sha2",
    "url": undefined,
  },
]
`;
